File: Bag3/Code/Stations/Icon Display Station/iconlib/emit.py
"""
emit.py -- write the final ICON tuple (device contract unchanged: a 16-line,
16-tuples-per-line `ICON = (...)` file, same shape icon_test.py/main.py
already read) and the --check lint pass.
"""

W = 16
H = 16
CH_FLOOR = 20          # nonzero authored channel below this truncates to
                        # <1 on-device at INTENSITY=0.30 -- see main.py scale()
MAX_LIT_COLORS = 8
MIN_FEATURE_CELLS = 4
SEPARATION_MIN = 25    # min max-channel delta between two adjacent distinct colors


def enforce_floor(pixels, floor=CH_FLOOR):
    """Any nonzero channel below `floor` is raised to `floor` -- keeps icon
    files robust to future INTENSITY changes instead of baking in today's
    value. Zero stays zero (that's an intentional off, not a dim color)."""
    out = []
    for r, g, b in pixels:
        out.append(tuple(c if c == 0 else max(floor, c) for c in (r, g, b)))
    return out


def write_icon(pixels, out_path, name_comment=None):
    with open(out_path, "w") as f:
        if name_comment:
            f.write("# %s\n" % name_comment)
        f.write("ICON = (\n")
        for row in range(H):
            f.write("    " + ", ".join(str(p) for p in pixels[row * W:(row + 1) * W]) + ",\n")
        f.write(")\n")


def lint(pixels, fills, decisions, cells_won, intensity):
    """
    Returns a list of human-readable warning/error strings. Empty list =
    clean. Checks (see design doc §Emit & lint):
      - unmapped fill above MIN_FRAC threshold (caller passes only mapped
        fills, so this is really "every fill decided")
      - a colored segment winning zero cells
      - more than MAX_LIT_COLORS distinct lit colors
      - a lit segment under MIN_FEATURE_CELLS
      - any lit channel that truncates to 0 on-device at `intensity`
      - two adjacent regions whose colors are within SEPARATION_MIN
    """
    problems = []

    lit_colors = set(p for p in pixels if p != (0, 0, 0))
    if len(lit_colors) > MAX_LIT_COLORS:
        problems.append("%d distinct lit colors on the grid (max recommended %d)"
                         % (len(lit_colors), MAX_LIT_COLORS))

    for i, d in enumerate(decisions):
        if d["role"] != "color":
            continue
        won = cells_won[i] if i < len(cells_won) else 0
        if won == 0:
            problems.append("segment %d (%s) is colored but won zero cells" % (i, fills[i][0]))
        elif won < MIN_FEATURE_CELLS:
            problems.append("segment %d (%s) only won %d cell(s) (min feature size %d)"
                             % (i, fills[i][0], won, MIN_FEATURE_CELLS))

    for r, g, b in pixels:
        for c in (r, g, b):
            if 0 < c and int(c * intensity) == 0:
                problems.append("channel value %d truncates to 0 on-device at intensity=%.2f" % (c, intensity))
                break

    for row in range(H):
        for col in range(W):
            idx = row * W + col
            c1 = pixels[idx]
            for dr, dc in ((0, 1), (1, 0)):
                r2, c2 = row + dr, col + dc
                if r2 >= H or c2 >= W:
                    continue
                c2v = pixels[r2 * W + c2]
                if c1 == (0, 0, 0) or c2v == (0, 0, 0) or c1 == c2v:
                    continue
                delta = max(abs(a - b) for a, b in zip(c1, c2v))
                if delta < SEPARATION_MIN:
                    problems.append(
                        "adjacent cells (%d,%d) %s and (%d,%d) %s differ by only %d"
                        % (row, col, c1, r2, c2, c2v, delta))

    # de-dupe truncation warnings (they repeat per-channel-value)
    seen = set()
    deduped = []
    for p in problems:
        if p not in seen:
            seen.add(p)
            deduped.append(p)
    return deduped

File: Bag3/Code/Stations/Icon Display Station/webapp/js/pipeline/emit.js
/**
 * emit.js -- floor enforcement, the device-contract ICON text writer,
 * ICON text, and the lint pass. iconText() must match write_icon()
 * character for character (see plan §Byte-exact emit) so a downloaded
 * icon diffs clean against the Python CLI's output.
 */

import { W, H, CH_FLOOR, MAX_LIT_COLORS, MIN_FEATURE_CELLS } from "./constants.js";
import { ledDeltaE, JND } from "./ledGamut.js";

/** Any nonzero channel below `floor` is raised to `floor`. Zero stays zero. */
export function enforceFloor(pixels, floor = CH_FLOOR) {
  return pixels.map(([r, g, b]) => [r, g, b].map((c) => (c === 0 ? 0 : Math.max(floor, c))));
}

/**
 * Byte-exact match for iconlib/emit.py:write_icon's text output (minus the
 * header comment, which the web app writes differently -- see webapp
 * README's parity notes).
 * @param {Array<[number,number,number]>} pixels  flat 256, row-major
 * @param {string} [nameComment]
 */
export function iconText(pixels, nameComment) {
  const lines = [];
  if (nameComment) lines.push(`# ${nameComment}`);
  lines.push("ICON = (");
  for (let row = 0; row < H; row++) {
    const rowPixels = pixels.slice(row * W, (row + 1) * W);
    const tuples = rowPixels.map(([r, g, b]) => `(${r}, ${g}, ${b})`);
    lines.push(`    ${tuples.join(", ")},`);
  }
  lines.push(")");
  return lines.join("\n") + "\n";
}

function pixEq(a, b) {
  return a[0] === b[0] && a[1] === b[1] && a[2] === b[2];
}
const isBlack = (p) => p[0] === 0 && p[1] === 0 && p[2] === 0;

/**
 * @returns {string[]} human-readable warnings, empty = clean.
 */
export function lint(pixels, fills, decisions, cellsWonArr, intensity) {
  const problems = [];

  const litKeys = new Set();
  for (const p of pixels) if (!isBlack(p)) litKeys.add(p.join(","));
  if (litKeys.size > MAX_LIT_COLORS) {
    problems.push(`${litKeys.size} distinct lit colors on the grid (max recommended ${MAX_LIT_COLORS})`);
  }

  for (let i = 0; i < decisions.length; i++) {
    const d = decisions[i];
    if (d.role !== "color") continue;
    const won = i < cellsWonArr.length ? cellsWonArr[i] : 0;
    const rgbStr = `(${fills[i].rgb.join(", ")})`;
    if (won === 0) {
      problems.push(`segment ${i} ${rgbStr} is colored but won zero cells`);
    } else if (won < MIN_FEATURE_CELLS) {
      problems.push(`segment ${i} ${rgbStr} only won ${won} cell(s) (min feature size ${MIN_FEATURE_CELLS})`);
    }
  }

  // Two segments authored to colours the panel cannot distinguish: the user
  // sees two colours in the editor and one colour on the hardware.
  for (let i = 0; i < decisions.length; i++) {
    for (let j = i + 1; j < decisions.length; j++) {
      const a = decisions[i];
      const b = decisions[j];
      if (a.role !== "color" || b.role !== "color" || !a.color || !b.color) continue;
      if ((cellsWonArr[i] ?? 0) === 0 || (cellsWonArr[j] ?? 0) === 0) continue;
      const dE = ledDeltaE(a.color, b.color, intensity);
      if (dE < JND) {
        problems.push(
          `segments ${i} and ${j} are different colours in the editor but render ` +
            `the same on the panel (dE ${dE.toFixed(3)}, need ${JND})`
        );
      }
    }
  }

  for (const [r, g, b] of pixels) {
    for (const c of [r, g, b]) {
      if (c > 0 && Math.trunc(c * intensity) === 0) {
        problems.push(`channel value ${c} truncates to 0 on-device at intensity=${intensity.toFixed(2)}`);
        break;
      }
    }
  }

  for (let row = 0; row < H; row++) {
    for (let col = 0; col < W; col++) {
      const idx = row * W + col;
      const c1 = pixels[idx];
      for (const [dr, dc] of [
        [0, 1],
        [1, 0],
      ]) {
        const r2 = row + dr;
        const c2 = col + dc;
        if (r2 >= H || c2 >= W) continue;
        const c2v = pixels[r2 * W + c2];
        if (isBlack(c1) || isBlack(c2v) || pixEq(c1, c2v)) continue;
        // Separation is measured on the PANEL, not on authored bytes. The
        // old max-channel-delta test compared duty values, which does not
        // answer "will a viewer tell these apart?" -- the device's per-channel
        // weighting and its intensity truncation both intervene. It missed
        // real collisions: a pale pink and a pale orange sitting 50 apart in
        // authored bytes are dE 0.035 on the panel, i.e. the same colour.
        const dE = ledDeltaE(c1, c2v, intensity);
        if (dE < JND) {
          problems.push(
            `adjacent cells (${row},${col}) (${c1.join(", ")}) and (${r2},${c2}) (${c2v.join(", ")}) ` +
              `look the same on the panel (dE ${dE.toFixed(3)}, need ${JND})`
          );
        }
      }
    }
  }

  const seen = new Set();
  const deduped = [];
  for (const p of problems) {
    if (!seen.has(p)) {
      seen.add(p);
      deduped.push(p);
    }
  }
  return deduped;
}
